tests/test_tw_csm2.py

```diff
@@ -0,0 +1,1943 @@
+        assert triggerable_withdrawals_gateway.hasRole(add_full_withdrawal_request_role, validators_exit_bus_oracle), "VEBO should have ADD_FULL_WITHDRAWAL_REQUEST_ROLE on TWG"
+
+        # Step 1.10: Validate DualGovernance tiebreaker connection
+        final_tiebreaker_details = dual_governance.getTiebreakerDetails()
+        final_tiebreakers = final_tiebreaker_details[3]  # sealableWithdrawalBlockers
+        assert TRIGGERABLE_WITHDRAWALS_GATEWAY in final_tiebreakers, "TWG should be in tiebreaker list after upgrade"
+
+        # Steps 1.11-1.12: Validate Withdrawal Vault upgrade
+        assert get_wv_contract_proxy_impl(withdrawal_vault) == WITHDRAWAL_VAULT_IMPL, "Withdrawal Vault implementation should be updated"
+        assert withdrawal_vault.getContractVersion() == 2, "Withdrawal Vault version should be 2 after finalizeUpgrade_v2"
+
+        # Steps 1.13-1.16: Validate Accounting Oracle upgrade
+        assert get_ossifiable_proxy_impl(accounting_oracle) == ACCOUNTING_ORACLE_IMPL, "Accounting Oracle implementation should be updated"
+        assert not accounting_oracle.hasRole(accounting_oracle.MANAGE_CONSENSUS_VERSION_ROLE(), agent), "Agent should not have MANAGE_CONSENSUS_VERSION_ROLE on AO"
+        assert accounting_oracle.getConsensusVersion() == AO_CONSENSUS_VERSION, f"AO consensus version should be set to {AO_CONSENSUS_VERSION}"
+
+        # Step 1.17: Validate AO finalizeUpgrade_v3
+        assert accounting_oracle.getContractVersion() == 3, "AO contract version should be 3 after finalizeUpgrade_v3"
+
+        # Steps 1.18-1.21: Validate Staking Router upgrade
+        assert get_ossifiable_proxy_impl(staking_router) == STAKING_ROUTER_IMPL, "Staking Router implementation should be updated"
+        assert staking_router.getContractVersion() == 3, "Staking Router version should be 3 after finalizeUpgrade_v3"
+        assert staking_router.hasRole(staking_router.REPORT_VALIDATOR_EXITING_STATUS_ROLE(), VALIDATOR_EXIT_VERIFIER), "ValidatorExitVerifier should have REPORT_VALIDATOR_EXITING_STATUS_ROLE on SR"
+        assert staking_router.hasRole(staking_router.REPORT_VALIDATOR_EXIT_TRIGGERED_ROLE(), triggerable_withdrawals_gateway), "TWG should have REPORT_VALIDATOR_EXIT_TRIGGERED_ROLE on SR"
+
+        # Steps 1.22-1.27: Validate NOR and sDVT updates
+        assert not acl.hasPermission(agent, kernel, app_manager_role), "Agent should not have APP_MANAGER_ROLE after vote"
+        assert nor.getContractVersion() == 4, "Node Operators Registry version should be updated to 4"
+        assert simple_dvt.getContractVersion() == 4, "Simple DVT version should be updated to 4"
+        assert nor.exitDeadlineThreshold(0) == NOR_EXIT_DEADLINE_IN_SEC, "NOR exit deadline threshold should be set correctly after finalizeUpgrade_v4"
+        assert simple_dvt.exitDeadlineThreshold(0) == NOR_EXIT_DEADLINE_IN_SEC, "sDVT exit deadline threshold should be set correctly after finalizeUpgrade_v4"
```

Comment:
Consider checking that `Proxy.implementation()` value updated correctly after upgrade